src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -257,91 +320,102 @@ public boolean isFingerprintArtifacts() {
     }
 
     @Override
-    public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener)
-            throws InterruptedException, IOException {
-        upgradeIfNecessary(build.getProject());
+    public void perform(@Nonnull Run<?, ?> build, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        if (build instanceof AbstractBuild) {
+            upgradeIfNecessary(((AbstractBuild)build).getProject());
+        }
+
+        EnvVars env;
+        if (build instanceof AbstractBuild) {
+            env = build.getEnvironment(listener);
```

Comment:
Wrong: [JENKINS-30357](https://issues.jenkins-ci.org/browse/JENKINS-30357) (should be lifted outside `if`-statement)